crypto/s2n_hash.c

```diff
@@ -169,13 +176,301 @@ int s2n_hash_digest(struct s2n_hash_state *state, void *out, uint32_t size)
+
+    int r;
+    switch (from->alg) {
+    case S2N_HASH_NONE:
+        r = 1;
+        break;
+    case S2N_HASH_MD5:
+    case S2N_HASH_SHA1:
+    case S2N_HASH_SHA224:
+    case S2N_HASH_SHA256:
+    case S2N_HASH_SHA384:
+    case S2N_HASH_SHA512:
+        r = EVP_MD_CTX_copy_ex(to_primary_ctx, from_primary_ctx);
+        break;
+    case S2N_HASH_MD5_SHA1:
+        r = EVP_MD_CTX_copy_ex(to_primary_ctx, from_primary_ctx);
+        r &= EVP_MD_CTX_copy_ex(to_md5_secondary_ctx, from_md5_secondary_ctx);
+        break;
+    default:
+        S2N_ERROR(S2N_ERR_HASH_INVALID_ALGORITHM);
+    }
+
+    if (r == 0) {
+        S2N_ERROR(S2N_ERR_HASH_COPY_FAILED);
+    }
+    to->hash_impl = from->hash_impl;
+    to->alg = from->alg;
+
+    return 0;
+}
+
+static int s2n_evp_hash_reset(struct s2n_hash_state *state)
+{
+    EVP_MD_CTX *primary_ctx = state->digest.high_level.evp.ctx;
+    EVP_MD_CTX *md5_secondary_ctx = state->digest.high_level.evp_md5_secondary.ctx;
+
+    int r;
+    r = S2N_EVP_MD_CTX_RESET(primary_ctx);
+    
+    if (state->alg == S2N_HASH_MD5_SHA1) {
+        r &= S2N_EVP_MD_CTX_RESET(md5_secondary_ctx);
+    }
+
+    if (r == 0) {
+        S2N_ERROR(S2N_ERR_HASH_WIPE_FAILED);
+    }
+
+    return s2n_evp_hash_init(state, state->alg);
+}
+
+static int s2n_evp_hash_free(struct s2n_hash_state *state)
+{
+    EVP_MD_CTX *primary_ctx = state->digest.high_level.evp.ctx;
+    EVP_MD_CTX *md5_secondary_ctx = state->digest.high_level.evp_md5_secondary.ctx;
+
+    S2N_EVP_MD_CTX_FREE(primary_ctx);
+    S2N_EVP_MD_CTX_FREE(md5_secondary_ctx);
+    primary_ctx = NULL;
+    md5_secondary_ctx = NULL;
+
+    return 0;
+}
+
+static const struct s2n_hash s2n_low_level_hash = {
+    .new = &s2n_low_level_hash_new,
+    .init = &s2n_low_level_hash_init,
+    .update = &s2n_low_level_hash_update,
+    .digest = &s2n_low_level_hash_digest,
+    .copy = &s2n_low_level_hash_copy,
+    .reset = &s2n_low_level_hash_reset,
+    .free = &s2n_low_level_hash_free,
+};
+
+static const struct s2n_hash s2n_evp_hash = {
+    .new = &s2n_evp_hash_new,
+    .init = &s2n_evp_hash_init,
+    .update = &s2n_evp_hash_update,
+    .digest = &s2n_evp_hash_digest,
+    .copy = &s2n_evp_hash_copy,
+    .reset = &s2n_evp_hash_reset,
+    .free = &s2n_evp_hash_free,
+};
+
+int s2n_hash_new(struct s2n_hash_state *state)
+{
+    /* Set hash_impl on initial hash creation.
+     * When in FIPS mode, the EVP API's must be used for hashes.
+     */
+    s2n_is_in_fips_mode() ? (state->hash_impl = &s2n_evp_hash) : (state->hash_impl = &s2n_low_level_hash);
+
+    return state->hash_impl->new(state);
+}
+
+int s2n_hash_init(struct s2n_hash_state *state, s2n_hash_algorithm alg)
+{
+    /* Ensure that hash_impl is set, as it may have been reset for s2n_hash_state on s2n_connection_wipe.
+     * When in FIPS mode, the EVP API's must be used for hashes.
+     */
+    s2n_is_in_fips_mode() ? (state->hash_impl = &s2n_evp_hash) : (state->hash_impl = &s2n_low_level_hash);
```

Comment:
Agreed - added `s2n_hash_set_impl` (moved 'set' to conform with other s2n_hash function names).